File: bot/data/xp_thresholds.py
"""D&D 5e XP thresholds per level.

XP_THRESHOLDS[i] is the minimum XP required to reach level i+1.
Index 0 → Level 1 (0 XP), index 19 → Level 20 (355 000 XP).
"""

XP_THRESHOLDS: tuple[int, ...] = (
    0,        # Level 1
    300,      # Level 2
    900,      # Level 3
    2_700,    # Level 4
    6_500,    # Level 5
    14_000,   # Level 6
    23_000,   # Level 7
    34_000,   # Level 8
    48_000,   # Level 9
    64_000,   # Level 10
    85_000,   # Level 11
    100_000,  # Level 12
    120_000,  # Level 13
    140_000,  # Level 14
    165_000,  # Level 15
    195_000,  # Level 16
    225_000,  # Level 17
    265_000,  # Level 18
    305_000,  # Level 19
    355_000,  # Level 20
)


def xp_to_level(xp: int) -> int:
    """Return the D&D 5e level corresponding to *xp* total experience points."""
    level = 1
    for threshold in XP_THRESHOLDS:
        if xp >= threshold:
            level = XP_THRESHOLDS.index(threshold) + 1
        else:
            break
    return level


def xp_for_next_level(current_xp: int) -> tuple[int, int | None]:
    """Return (current_level, xp_needed_for_next_level).

    xp_needed_for_next_level is None when the character is already level 20.
    """
    current_level = xp_to_level(current_xp)
    if current_level >= 20:
        return 20, None
    next_threshold = XP_THRESHOLDS[current_level]  # index = level (0-based → next level)
    return current_level, next_threshold - current_xp

File: webapp/src/lib/xpThresholds.ts
/**
 * D&D 5e experience-point thresholds (SRD).
 * Index `i` = XP required to reach level `i + 1`.
 * Index 0 (= 0) is the baseline for level 1.
 */
export const XP_THRESHOLDS: readonly number[] = [
  0,
  300, 900, 2700, 6500, 14000,
  23000, 34000, 48000, 64000, 85000,
  100000, 120000, 140000, 165000, 195000,
  225000, 265000, 305000, 355000,
] as const

/** Derive the current character level from accumulated XP (capped at 20). */
export function levelFromXp(xp: number): number {
  let level = 1
  for (let i = 1; i < XP_THRESHOLDS.length; i++) {
    if (xp >= XP_THRESHOLDS[i]) level = i + 1
    else break
  }
  return Math.min(level, 20)
}

/** XP threshold needed to reach `currentLevel + 1`. Null if at cap. */
export function getNextLevelThreshold(currentLevel: number): number | null {
  return XP_THRESHOLDS[currentLevel] ?? null
}

/**
 * Quick-XP button amounts, proportional to XP remaining until the next level.
 *
 * Formula: 2% / 7% / 20% / 50% of `xpToNext`, rounded to multiples of 10,
 * with a minimum of 5 XP per button; adjacent duplicates after rounding are
 * removed so bottom-range progressions (e.g. xpToNext=50) collapse gracefully.
 *
 * Returns an empty array if `xpToNext <= 0` (character is at max level).
 */
export function quickXpAmounts(xpToNext: number): number[] {
  if (xpToNext <= 0) return []
  const PCTS = [0.02, 0.07, 0.20, 0.50] as const
  const MIN_AMOUNT = 5
  const raw = PCTS.map((p) => Math.max(MIN_AMOUNT, Math.round((p * xpToNext) / 10) * 10))
  return raw.filter((v, i) => i === 0 || v !== raw[i - 1])
}
